lib/ansible/plugins/action/enos.py

```diff
@@ -79,5 +83,18 @@ def run(self, tmp=None, task_vars=None):
 
             task_vars['ansible_socket'] = socket_path
 
+        # make sure we are in the right cli context which should be
+        # enable mode and not config module or exec mode
+        if socket_path is None:
+            socket_path = self._connection.socket_path
+
+        conn = Connection(socket_path)
+        out = conn.get_prompt()
+        if to_text(out, errors='surrogate_then_replace').strip().endswith(')#'):
+            display.vvvv('In Config mode, sending exit to device', self._play_context.remote_addr)
+            conn.send_command('exit')
```

Comment:
If you have an option of `end` command (similar to IOSXR) or and `abort` command (this will take care of any stale uncommitted changes). 

Also, when you exit out of configure mode, I believe you will be in `exec` mode? In that case, you might to `enable` command every time and not just in the `else` part?
OR even better (however make sure you are in exec mode before doing this)

if to_text(out, errors='surrogate_then_replace').strip().endswith('>'):
    conn.send_command('enable')

Reply:
Hi Kedar,

Thank you very much for the valuable feedback. 
Now let me explain my logic here. There could be three scenarios here. 
Case 1. First time login. Unfortunately this will be the response an ENOS system.

login as: admin
Using keyboard-interactive authentication.
Enter login password:
Lenovo RackSwitch G8272.
------------------------------------------------------------------
Note: The current running configuration includes changes
      that have NOT been saved to FLASH.
------------------------------------------------------------------
RS G8272>
Nov 16 16:58:39 192.168.49.50 NOTICE  mgmt: admin(admin) login from host 10.38.100.104

This means that the end character is not ">" rather it cannot be predicted. But ">" is present there which I am using in Regular expression

Case 2  . Connection is already in enable mode.  In that case we need not do anything
Case 3 . Connection is already in configure terminal mode. In that the end character will be as shown below.
RS G8272#conf t
Enter configuration commands, one per line.  End with Ctrl/Z.
RS G8272(config)#

So my logic is like this. If its in configure mode, it has to come into enable mode , 
For that I will issue an exit command
Else if it is exec mode, it has to come to enable mode, for that I will issue an "enable" command
Else if its already in enable mode, another enable command will do nothing.

This is the logic I have coded there. I feel this logic is OK.  I am just taking care of all scenarios of a persistent connection. 
Behavior of ENOS switches are like this and I have to live with it.

Reply:
sounds good.